dpnp/backend/tests/test_broadcast_iterator.cpp

```diff
@@ -0,0 +1,202 @@
+            << ", result=" << data.result << ")";
+
+        return out;
+    }
+};
+
+class IteratorBroadcasting : public ::testing::TestWithParam<IteratorParameters>
+{
+};
+
+TEST_P(IteratorBroadcasting, loop_broadcast)
+{
+    const IteratorParameters& param = GetParam();
+    const dpnpc_index_t result_size = param.result.size();
+
+    vector<dpnpc_value_t> input_data = get_input_data<dpnpc_value_t>(param.input_shape);
+    DPNPC_id<dpnpc_value_t> input(input_data.data(), param.input_shape);
+    input.broadcast_to_shape(param.output_shape);
+
+    ASSERT_EQ(input.get_output_size(), result_size);
+
+    vector<dpnpc_value_t> test_result(result_size, 42);
+    for (dpnpc_index_t output_id = 0; output_id < result_size; ++output_id)
+    {
+        test_result[output_id] = input[output_id];
+        std::cout << "test_result[" << output_id << "] = " << test_result[output_id] << std::endl;
+        std::cout << "param.result.at(" << output_id << ") = " << param.result.at(output_id) << std::endl;
+        EXPECT_EQ(test_result[output_id], param.result.at(output_id));
+    }
+}
+
+TEST_P(IteratorBroadcasting, sycl_broadcast)
+{
+    using data_type = double;
+
+    const IteratorParameters& param = GetParam();
+    const dpnpc_index_t result_size = param.result.size();
+    vector<data_type> result(result_size, 42);
+    data_type* result_ptr = result.data();
+
+    vector<data_type> input_data = get_input_data<data_type>(param.input_shape);
+    DPNPC_id<data_type> input(input_data.data(), param.input_shape);
+    input.broadcast_to_shape(param.output_shape);
+
+    ASSERT_EQ(input.get_output_size(), result_size);
+
+    cl::sycl::range<1> gws(result_size);
+    const DPNPC_id<data_type>* input_it = &input;
+    auto kernel_parallel_for_func = [=](cl::sycl::id<1> global_id) {
+        const size_t idx = global_id[0];
+        result_ptr[idx] = (*input_it)[idx];
+    };
+
+    auto kernel_func = [&](cl::sycl::handler& cgh) {
+        cgh.parallel_for<class test_sycl_reduce_axis_kernel>(gws, kernel_parallel_for_func);
+    };
+
+    cl::sycl::event event = DPNP_QUEUE.submit(kernel_func);
+    event.wait();
+
+    for (dpnpc_index_t i = 0; i < result_size; ++i)
+    {
+        EXPECT_EQ(result.at(i), param.result.at(i));
+    }
+}
+
+/**
+ * Expected values produced by following script:
+ *
+ * import numpy as np
+ *
+ * input_size = 12
+ * input_shape = [3, 4]
+ * input = np.arange(1, input_size + 1, dtype=np.int64).reshape(input_shape)
+ * print(f"input shape={input.shape}")
+ * print(f"input:\n{input}\n")
+ * 
+ * output_shape = [2, 3, 4]
+ * output = np.ones(output_shape, dtype=np.int64)
+ * print(f"output shape={output.shape}")
+ *
+ * result = input * output
+ * print(f"result={np.array2string(result.reshape(result.size), separator=', ')}\n", sep=", ")
+ */
+INSTANTIATE_TEST_SUITE_P(
+    TestBroadcastIterator,
+    IteratorBroadcasting,
+    testing::Values(
```

Comment:
👍 
Do you think it is good to add some combinations like "{1, 0}. {0}"?

Reply:
I wanted to add something like this `IteratorParameters{{2, 0, 1}, {2, 0, 4}, {}}));` as test parameter, but in this case we have empty result. So SYCL kernel doesn't make sense here, because we iterate by result size that is equal to 0 in this case. In real case we should return from the function before submitting such kernel to the queue. Moreover we can see below issue on GPU when submitting such kernel to the queue
`C++ exception with description "Native API failed. Native API returns: -30 (CL_INVALID_VALUE) -30 (CL_INVALID_VALUE)" thrown in the test body.`
That is why 2 tests on reduction fail https://github.com/IntelPython/dpnp/pull/661#issuecomment-810387036.